### PR TITLE
Fix dev dependency exclusion bug

### LINKED ISSUE #3827: Excluding dev dependencies throws Error
# This is a Bug Report

## Description

When packaging a serverless project, the dev dependency exclusion does not happen. It throws an Error that is caught with no action. 

The `excludeNodeDevDependencies` function in `/lib/plugins/package/lib/zipService.js` throws when it calls:

```js
const prodDependencies = childProcess
        .execSync('npm ls --prod=true --parseable=true --silent', {
          maxBuffer: 2000*1024
        })
        .toString().trim();
```

When catching the thrown error, it seems that the command is partially run, as there are some directories listed in `stdout`, but they stop in the middle. The `error` property is null, but the `status` is `1`. Here's a stack trace:

```sh
Error: Command failed: npm ls --prod=true --parseable=true --silent

    at checkExecSyncError (child_process.js:481:13)
    at Object.execSync (child_process.js:521:13)
    at _.forEach (/Users/rylewis/Documents/Dev/Repo/serverless/lib/plugins/package/lib/zipService.js:145:10)
    at arrayEach (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/lodash/lodash.js:537:11)
    at Function.forEach (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/lodash/lodash.js:9359:14)
    at excludeNodeDevDependencies (/Users/rylewis/Documents/Dev/Repo/serverless/lib/plugins/package/lib/zipService.js:136:7)
    at Package.excludeDevDependencies (/Users/rylewis/Documents/Dev/Repo/serverless/lib/plugins/package/lib/zipService.js:30:25)
    at Package.tryCatcher (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/rylewis/Documents/Dev/Repo/serverless/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
```

I've tried increasing the `timeout` and `maxBuffer` options, but no change. Running the command `npm ls --prod=true --parseable=true --silent` in my shell works correctly and returns all the directories.

## Additional Data

* ***Serverless Framework Version you're using***: 1.16.0
* ***Operating System***: macOS

### LINKED ISSUE #3869: Change in package/exclude usage?
Have there been changes in how to exclude paths from being included in the package? I see different archive sizes (*22kb* compared to *6.7mb*) when using `1.15.3` and `1.16.1`:

## Config

```yaml
package:
  exclude:
    - node_modules/**
```

Based on the [docs](https://serverless.com/framework/docs/providers/google/guide/packaging/#exclude--include) it should work like this.

## Using 1.15.3

Archive is ~22kb.

```bash
$ > yarn add serverless@1.15.3
$ > $(npm bin)/sls version && \
    $(npm bin)/sls package && \
    ls -ahl .serverless/sls-example.zip

1.15.3

Serverless: Packaging service...

-rw-r--r--  1 sbstjn  staff    22K Jun 27 21:58 .serverless/sls-example.zip
```

## Using 1.16.1

Archive is ~7mb

```bash
$ > yarn add serverless@1.16.1
$ > $(npm bin)/sls version && \
    $(npm bin)/sls package && \
    ls -ahl .serverless/sls-example.zip

1.16.1

Serverless: Packaging service...

-rw-r--r--  1 sbstjn  staff   6.7M Jun 27 22:00 .serverless/sls-example.zip
```

## NPM Dependencies

After moving the `serverless` package from `dependencies` to `devDependencies` in my `package.json`, the archive has `22kb` using `1.16.1` as well.

### 22kb

```json
{
  "devDependencies": {
    "serverless": "1.16.1"
  }
}
```

### 7mb

```json
{
  "dependencies": {
    "serverless": "1.16.1"
  }
}
```

### Expectation

I expected it does not matter if packages are in `dependencies` or `devDependecies` when using `package/exclude` in my configuration to exclude the `node_modules` folder.

### LINKED ISSUE #3874: Fix dev dependency exclusion bug when too many node modules are used
# This is a Bug Report

## Description

Right now we do a glob exclude of the whole `node_modules` directory and include only prod dependencies by using `npm ls` to identify them and a glob to re-include them.

However there's a bug with the maximum number of characters the buffer can hold.

Furthermore we should switch the behavior. We should list the dev dependencies and exclude them (while still including the whole `node_modules` directory) to ensure that the whole `node_modules` directory is still in included. Even if the command fails.

---

**Quick fix / other addition**

Add a config parameter which allows people to opt out of this feature.

---

Related #3827

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #3827
Closes #3874
Closes #3869

Fixes a bug where the dev dependency exclusion errors out due to a large number of dependencies (and string capacity limitations).

Also fixes the error where exclusion is not working anymore.

Speed is increased due to async operations ran concurrently.

## How did you implement it:

1. Switch from sync to async calls and compute them in a `Bluebird Promise.map`
2. Switched from `execSync` to `spawn` to get a stream (mitigates the buffer limit `execSync` introduces where the packages for exclusion are capped)
3. Switch from `npm ls --prod=true` to `npm ls --dev=true` so that we exclude only the `dev` dependencies rather than excluding all `node_modules` and then re-including the prod dependencies. This should avoid side-effects when the exclusion of dev dependencies fails.

## How can we verify it:

Take the steps described here: https://github.com/serverless/serverless/pull/3737#issue-233860537

## Todos:

- [x] Write tests
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pmuens** [comment]
Just a quick heads up. I investigated futher into this issue today and found out that we can use a solution which won't introduce an additional temp file.

[It looks like](https://stackoverflow.com/a/26995407) using `spawn` could solve the buffer overflow problem since it returns a stream we can then react on.

I'll update this PR the upcoming days with a solution for this.

**Comment by pmuens** [comment]
@ryanmurakami @cjelger @arabold @sbstjn @ajkerr I just pushed a fix for the issues with the current dev dependency exclusion.

The changes which were applied can be found in the [PR description](https://github.com/serverless/serverless/pull/3889#issue-240183777).

Would be nice if you could take this for a test drive and see if it resolves all the issues you brought up. This way we can get this PR in a mergeable state and publish it in a patch release the upcoming days.

Thanks in advance 👍

**Comment by ajkerr** [comment]
@pmuens The PR fixes the issues we were seeing with our project. Thanks!

**Comment by pmuens** [comment]
> @pmuens The PR fixes the issues we were seeing with our project. Thanks!

Great! 🎉  Thanks for confirming @ajkerr 👍

File: lib/plugins/package/lib/zipService.js [hunk 1]
```diff
@@ -131,53 +137,62 @@ function excludeNodeDevDependencies(servicePath) {
...
+            .filter((item) => item.length > 0);
+          return resolve(result);
+        });
 
-      if (includePatterns.length) {
-        // at first exclude the whole node_modules directory
-        // after that re-include the production relevant modules
-        exclude = _.union(exclude, [`${pathToDep}node_modules/**`]);
-        include = _.union(include, includePatterns);
+        childProc.stdout.on('error', () => resolve(result));
```
Comment (arabold) [inline lib/plugins/package/lib/zipService.js]: Wouldn't it be reasonable to abort in case of error and leave the exclude and includes untouched?

File: lib/plugins/package/lib/zipService.js [hunk 2]
```diff
@@ -114,9 +119,10 @@ module.exports = {
 };
 
 function excludeNodeDevDependencies(servicePath) {
-  const cwd = process.cwd();
-  let exclude = [];
-  let include = [];
+  const exAndIn = {
+    include: [],
+    exlucde: [],
```
Comment (ryanmurakami) [inline lib/plugins/package/lib/zipService.js]: typo: `exclude`

Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Woops! Nice catch @ryanmurakami 👍

File: lib/plugins/package/lib/zipService.js (hunk 1 above)
Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Yes, that sounds like good idea 👍

**Comment by ryanmurakami** [comment]
@pmuens Same result but worse. :-/ Since you changed to looking at dev dependencies, the output stops and one of the lines ends up excluding the entire `node_modules` folder. Smaller package size for sure. :-)

**Comment by ryanmurakami** [comment]
I'm going to mess around with using a temp file, see if I can make any progress on that.

**Comment by HyperBrain** [comment]
@pmuens Didn't you have a temp file version before? Maybe reverting to that would fix @ryanmurakami 's issue.

**Comment by ryanmurakami** [comment]
@pmuens @HyperBrain So, it looks like using a tempfile fixes the issue. 
I tried two methods. The first, hooking up a file write stream to the child process's `stdout` doesn't work because it still dies at 8kb.
The second is to just add `> deps.txt` to the end of the command. This works and sends everything to a file. Then I can just read from the file.

I can push that code, but there are a couple of concerns. The `package` command, with this approach, takes almost 2 minutes to complete, which is way too long. I switched the `ls` to look at production dependencies instead, and it cut the time in half. Oddly, the filesize was slightly larger. 

Anyways, I pushed my changes to my repo in this commit: https://github.com/ryanmurakami/serverless/commit/f137fbc92b21ad22f4591fd2896d51a11a9dd1c6

I can clean this up if you guys think it's a good approach.

**Comment by pmuens** [comment]
🤔 thanks for looking into this @ryanmurakami 👍 

@HyperBrain yes, we had this file based approach (originally brought up by @arabold). It would be nice if we could the file creation and rely on streams (since IO could be unreliable and slow).

We switched from `exec` to `spawn` since `spawn` returns a stream you can tap into whereas `exec` has a buffer with a fixed size. Maybe I'm missing something here. Isn't there any way to stream the data from `stdout` to avoid this buffer size limitation `exec` introduces?

I'll take a look into your commit @ryanmurakami tomorrow. Would be really nice if we could fix this issue w/o introducing an external file.

**Comment by ryanmurakami** [comment]
@pmuens I did a little more work and refactored the code so it's a little easier to riff on different approaches: https://github.com/ryanmurakami/serverless/tree/fix-dev-dependency-exclusion-bug

You can modify the `getDependenciesStream` and `getDependenciesFile`, then comment/uncomment lines 206 & 207 to switch each approach.

It appears this is some type of buffer paging that macOS is doing on the stdout stream. I don't see any differences when using `spawn` or `exec` in terms of the output changing. Both are cutting off at around 8kb whether you're piping `stdout` or just reading it after the fact.

I haven't been able to think of another approach except for a temp file. The only other thing is if there was some way we could emulate a file but actually write it directly in memory. I'm not even sure that's possible though.

**Comment by ryanmurakami** [comment]
Also, thanks so much for working hard on this @pmuens 🙏🙏🙏

**Comment by arabold** [comment]
The reason both `exec` and `spawn` fail to grab the whole output lies in `npm` itself. `npm` calls a `process.exit()` _before_ it has flushed its `stdout`. That exit signal is catched by `exec` and `spawn` which stops reading the buffer immediately (which happens to be after 8k). I don't have the respective issue in npm at hand right now. It was linked in my original commit IIRC.

Long story short: From what I understand only a temp file will fully solve this problem.

**Comment by ajkerr** [comment]
I believe that npm has purged most of their open GitHub bugs (for better or worse), and are only working on npm5+. Given that Node's current LTS is 6.x with npm 3.x, I don't think waiting for an npm fix is a great option.

**Comment by pmuens** [comment]
Thanks for the in-depth discussion about this 👍 

I found @arabold research comment and will link it here: https://github.com/serverless/serverless/pull/3737#issuecomment-312298152 (thanks for looking into this @arabold 👍).

Seems like there's no way around a file-based approach.

Next up I'll take a look @ryanmurakami implementation and update this PR so that we hopefully can fix this soon (and release a patch).

Again. Thanks for all the work and discussions on this here 💯!

**Comment by pmuens** [comment]
I just looked through @ryanmurakami branch (thanks for putting this together!) and pushed an update and switched back to the tmp file approach. I kept all the operations async.

Works on my machine. Would be nice if you could try it and see if it fixes the problem.

Another question: Is the exec command (`>> foo.file`) OS independent and works on all Windows machines? 🤔 

/cc @ryanmurakami @arabold

File: lib/plugins/package/lib/zipService.js [hunk 3]
```diff
@@ -131,53 +145,39 @@ function excludeNodeDevDependencies(servicePath) {
...
+      const dirWithPackageJson = fullPath.replace(path.join(path.sep, 'package.json'), '');
+
+      return childProcess.execAsync(
+        `npm ls --dev=true --parseable=true --silent >> ${nodeDevDepFile}`,
+        { cwd: dirWithPackageJson }
+      );
+    })
+    .then(() => fs.readFileAsync(nodeDevDepFile))
+    .then((fileContent) => {
+      const dependencies = fileContent.toString('utf8').split('\n');
```
Comment (HyperBrain) [inline lib/plugins/package/lib/zipService.js]: Shouldn't this be deduplicated (as it might be a combination of multiple package.json files and could contain duplicates)?
```
const dependencies = _.uniq(_.split(fileContent.toString('utf8'), '\n'));
```

I also suggest to use lodash, as otherwise you have to test here that fileContent is not nil, before using toString().

File: lib/plugins/package/lib/zipService.js [hunk 4]
```diff
@@ -131,53 +145,39 @@ function excludeNodeDevDependencies(servicePath) {
...
 
-      const includePatterns = _.map(prodDependencyPaths, (depPath) =>
-        `${pathToDep}${depPath}/**`);
-
-      if (includePatterns.length) {
-        // at first exclude the whole node_modules directory
-        // after that re-include the production relevant modules
-        exclude = _.union(exclude, [`${pathToDep}node_modules/**`]);
-        include = _.union(include, includePatterns);
+      if (dependencies.length) {
```
Comment (HyperBrain) [inline lib/plugins/package/lib/zipService.js]:
```
if (!_.isEmpty(dependencies))
...
```

File: lib/plugins/package/lib/zipService.js (hunk 3 above)
Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Good catch @HyperBrain 👍 I'll update it so that it uses Lodash.

File: lib/plugins/package/lib/zipService.js [hunk 5]
```diff
@@ -2,12 +2,15 @@
 
 /* eslint-disable no-use-before-define */
 /* eslint-disable no-param-reassign */
+/* eslint-disable consistent-return */
```
Comment (HyperBrain) [inline lib/plugins/package/lib/zipService.js]: Which lines are affected by this? Can the implementation be changed to comply to the rule, or is there a specific reason, why this cannot be done?

Comment (HyperBrain) [inline lib/plugins/package/lib/zipService.js]: Addition: Imo it's generally bad to disable linter check per file. It would be better to put an `elslint-disable-line consistent-return` at the specific location and comment why this rule is needed to be overridden there.

File: lib/plugins/package/lib/zipService.js [hunk 6]
```diff
@@ -142,47 +150,24 @@ function excludeNodeDevDependencies(servicePath) {
...
+    // NOTE: using mapSeries here for a sequential computation (w/o race conditions)
+    return BbPromise.mapSeries(packageJsonPaths, (packageJsonPath) => {
       // the path where the package.json file lives
       const fullPath = path.join(servicePath, packageJsonPath);
       const dirWithPackageJson = fullPath.replace(path.join(path.sep, 'package.json'), '');
 
-      const childProc = childProcess.spawn('npm',
-        ['ls', '--dev=true', '--parseable=true', '--silent'],
+      return childProcess.execAsync(
+        `npm ls --dev=true --parseable=true --silent >> ${nodeDevDepFile}`,
```
Comment (arabold) [inline lib/plugins/package/lib/zipService.js]: If you append to `${nodeDevDepFile}` then you should make sure to delete it before the `BbPromise.mapSeries()` above . Otherwise it might have some surprises in store for you. Why not put it in the artifacts folder (i.e. `.serverless` or whatever is provided in the `--package` command line) though? I think that folder is cleaned and recreated on every packaging?

Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: 🤔 I thought `mapSeries` runs sequential. Maybe I was missing something there.

Right now the name of the `nodeDevDepFile` is generated with a hash in the name and stored in the OS tempDir. Putting it in `.serverless` could also be an option though.

Comment (arabold) [inline lib/plugins/package/lib/zipService.js]: My concern is if there's a leftover from a previous run. Thus the file already exists before serverless is started.

Comment (arabold) [inline lib/plugins/package/lib/zipService.js]: The random number you're using does certainly help though. Didn't notice this before.

**Comment by ryanmurakami** [comment]
@pmuens I've been hacking on this new solution. The temp file is working good, but `execAsync` is rejecting and causing no dependencies to be removed. If you revert it to sync and wrap it in a `try / catch`, then everything works. It's super messy, I know. 

I was able to get a repo that repros the issue. You'll need macOS to see it, but using the `package.json` in [this repo](https://github.com/ryanmurakami/hella-deps) will generate ~1000 dev dependencies. Running `serverless package` with the current PR will create a zip file of ~150mb without excluding any dependencies.

**Comment by ryanmurakami** [comment]
@pmuens It does appear that `>>` is OS independent. I can try this out on a windows pc at home.

File: lib/plugins/package/lib/zipService.js (hunk 5 above)
Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Thanks @HyperBrain 👍 Yes, that's a good suggestion. I'll add the comments to the lines.

File: lib/plugins/package/lib/zipService.js (hunk 6 above)
Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Alright gotcha. Yes, that's a good catch 👍 .

I re-thought about keeping it in `.serverless` however IMHO keeping it in a tmpdir which is cleaned up by the OS is slightly better since the `.serverless` dir could change (although we have utils to get the path) and is used to store deployment artifacts.

**Comment by pmuens** [comment]
Thanks for testing it our @ryanmurakami 👍 Really appreciate your help here (since I'm running it with Linux where everything is fine).

> @pmuens I've been hacking on this new solution. The temp file is working good, but execAsync is rejecting and causing no dependencies to be removed. If you revert it to sync and wrap it in a try / catch, then everything works. It's super messy, I know.

🤔 that's interesting. Do you know why it's rejected? So is there an implementation problem with the Async approach or is this also related to the Mac OS buffer issue we're trying to solve here?

*(Would be super nice if we could implement this in an async way to keep things as fast as they could be)*

> I was able to get a repo that repros the issue. You'll need macOS to see it, but using the package.json in this repo will generate ~1000 dev dependencies. Running serverless package with the current PR will create a zip file of ~150mb without excluding any dependencies.

Thanks for sharing that. I'll look into it ASAP!

**Comment by ryanmurakami** [comment]
@pmuens It looks like the command passed to exec is returning a 1 code, so the exec call is throwing an error. As an async promise, it ends up rejecting and the current behavior just resolves an empty dependencies array to exclude. There's probably a way to still do it async but still check the file if it rejects. I can take a stab at it tomorrow, unless you have some ideas. 

I think we're almost there! Thanks again for spending so much time on this issue!

**Comment by HyperBrain** [comment]
@ryanmurakami Can you check somehow WHY it returns 1 - more specifically, if the "npm" or the shell (">> filename") errors ? My suspicion would be that, because we append to the file, it might be locked if more than one npm is executed in parallel.

It also seems useful, if we could get the `stderr` stream from the exec callback and check if it's empty to have the ability to print some of the errors. But there's a problem, that the callback is called with 3 (!!) and not 2 parameters:
https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback

So the call itself could be written so:
```
return BbPromise.fromCallback(callback => {
  childProcess.exec(
    `npm ls --dev=true --parseable=true --silent >> ${nodeDevDepFile}`,
    { cwd: dirWithPackageJson },
    (err, stdout, stderr) => {
      if (err) {
        // Log error details if exec failed
        this.serverless.cli.log(`npm exec error: ${err.code}`);
        this.serverless.cli.log(`Error output: ${stderr}`);
      }
      callback(err);
    }
  );
});
```

This implementation makes sure that the exec call is tracked corretly and in case of any errors, the reson is shown.

**Comment by pmuens** [comment]
> @pmuens It looks like the command passed to exec is returning a 1 code, so the exec call is throwing an error. As an async promise, it ends up rejecting and the current behavior just resolves an empty dependencies array to exclude. There's probably a way to still do it async but still check the file if it rejects. I can take a stab at it tomorrow, unless you have some ideas.

Thanks for getting back and looking into it @ryanmurakami 👍 That sounds reasonable (but still unsure why it returns a `1`). Would be nice if you could look at it and we can reflect the updates here.

I'll also take a look at it the upcoming days. Can't wait to get this fixed and ship it in a patch release!

Good ideas @HyperBrain 👍 Thanks for taking the time to troubleshoot it.

On Linux it failed if you do an `npm ls` in a directory where a package.json is available but no `node_modules` directory is present (as far as I remember). Not 100% sure though.

BTW. Here's smth. interesting I found on exit code for `npm ls`:

- https://github.com/npm/npm/issues/11160
- https://github.com/npm/npm/issues/2798

That could be the reason why it fails 😬

**Comment by pmuens** [comment]
I just updated the PR according to @HyperBrain comments to use the `lodash` utilites (thanks for that 👍).

Furthermore I replace the `exec` logic with the code snippet by @HyperBrain and ran it against a service which has a package.json file, but no `npm` modules installed yet (so no `node_modules` dir). I figured that `npm` returns a 1 for the exit code in that case (and fails).

To fix it I added a `.catch()` at the end of the `execAsync` call. @ryanmurakami this should do the same as you've proposed with switching it back to sync and wrapping it in a `try` / `catch` block.

This fixes the problem of interrupted exclusions on my machine.

Would be nice if this can be re-tested so that we can get this into a mergeable state and ship it soon 👍 . Thanks in advance!

**Comment by pmuens** [comment]
Here's a test service which can be used to verify that the package exclusion works correctly.

It consists of 3 different directory levels where each level has a respective `package.json` file. Just run `npm install` in the different directories to install the dependencies and dev dependencies.

You can also e.g. only run `npm install` in the root level to force an exit code 1 (since `npm ls` in a directory with no installed modules will fail).

/cc @eahefnawy @ryanmurakami @HyperBrain @arabold 

[dev-dependency-exclusion.zip](https://github.com/serverless/serverless/files/1153238/dev-dependency-exclusion.zip)

**Comment by ryanmurakami** [comment]
@pmuens Thanks again for your continued work on this!
Testing it on my problem directory and this seems to have solved the issue.

I'm encountering something different, but I think it has to do with how I'm testing and not the actual code. I'm testing the PR by doing an `npm link serverless` and it appears that the zip process is zipping up the entire serverless directory with the package. Any idea why this is happening or a better way to test the new code? Regardless, I think you've got it with the most recent changes and it should be good to go.

**Comment by ryanmurakami** [comment]
Okay, I used `npx` and was able to validate this fix working good. 

On a side note, it takes almost 3x longer to package my large project with this fix. I noticed the time is mostly spent after the dependencies are all resolved, most likely when `archiver` is zipping directories. Not a huge deal considering I think my project is an outlier, but just something I wanted to mention.

**Comment by pmuens** [comment]
> Okay, I used npx and was able to validate this fix working good.

Nice! Thanks for testing it out @ryanmurakami 👍 Great to hear that it finally works on your machine!

> On a side note, it takes almost 3x longer to package my large project with this fix. I noticed the time is mostly spent after the dependencies are all resolved, most likely when archiver is zipping directories. Not a huge deal considering I think my project is an outlier, but just something I wanted to mention.

🤔 That's definitely smth. we should look into next up. Thanks for bringing this to our attention! 👍 

---

I pushed some minor changes and fixed / add a couple of tests. This PR is GTM from my side.

/cc @eahefnawy

File: lib/plugins/package/lib/zipService.js [hunk 7]
```diff
@@ -164,14 +168,17 @@ function excludeNodeDevDependencies(servicePath) {
     })
     .then(() => fs.readFileAsync(nodeDevDepFile))
     .then((fileContent) => {
-      const dependencies = _.uniq(_.split(fileContent.toString('utf8'), '\n'));
+      const dependencies = _.filter(
```
Comment (ryanmurakami) [inline lib/plugins/package/lib/zipService.js]: Could you use `_.compact` here instead of `_.filter`?
https://lodash.com/docs/4.17.4#compact

File: lib/plugins/package/lib/zipService.js [hunk 8]
```diff
@@ -131,53 +144,48 @@ function excludeNodeDevDependencies(servicePath) {
...
+      // we added a catch which resolves so that npm commands with an exit code of 1
+      // (e.g. if the package.json is invalid) won't crash the dev dependency exclusion process
+      return childProcess.execAsync(
+        `npm ls --dev=true --parseable=true --silent >> ${nodeDevDepFile}`,
+        { cwd: dirWithPackageJson }
+      ).catch(() => BbPromise.resolve());
+    })
+    .then(() => fs.readFileAsync(nodeDevDepFile))
+    .then((fileContent) => {
+      const dependencies = _.filter(
```
Comment (ryanmurakami) [inline lib/plugins/package/lib/zipService.js]: Could you use _.compact here instead of _.filter?
https://lodash.com/docs/4.17.4#compact

Comment (pmuens) [inline lib/plugins/package/lib/zipService.js]: Thanks. That's a good idea @ryanmurakami 👍

**Comment by eahefnawy** [comment]
@pmuens following your instructions, I was able to validate that exclusion works as expected! 👍

G2M! 😊

**Comment by sasidharkanumuri** [comment]
I deployed https://github.com/react-boilerplate/react-boilerplate.git over lambda in AWS by cloudformation and I wanted to test this app now by lambda function , any ideas on how to test its localhost

**Comment by pmuens** [comment]
@sasidharkanumuri thanks for commenting 👍

I think the best place to get help for such problems would be the [Serverless Gitter](http://gitter.im/serverless/serverless) of our [Forum](http://forum.serverless.com).

Let us know if you need any help with that 👍

**Comment by sasidharkanumuri** [comment]
Thanks @pmuens I just posted my question in the Serverless Gitter forum

**Comment by pmuens** [comment]
> Thanks @pmuens I just posted my question in the Serverless Gitter forum

Great! Thanks for that 👍 Hopefully your problem can be resolved soon!

**Comment by andythecoderman** [comment]
I'm still getting the dev dependencies, or at least most of them using serverless v1.23.0 and serverless-plugin-typescript v1.1.3 

Do I need a later version to get this branch?

**Comment by tommedema** [comment]
I can still reproduce this issue. See my example mono repo here:

https://github.com/tommedema/serverless-mono-example

Steps to reproduce:

- `cd packages/serverless-random`
- `yarn run package`
- note that the package is 22MB with a lot of dev dependencies, while it should be < 1 MB

**Comment by tommedema** [comment]
I'm facing a few issues related to this. I've created a sample repo here:

https://github.com/tommedema/serverless-mono-example

Steps to reproduce
- `yarn install`
- `cd packages/serverless-random`
- `yarn package`

Note that the package now has a few presumably dev dependencies which shouldn't be there:

- ts-node
- make-error
- arrify
- yn
- .bin/ts-node

Also, the bundle has redundant source files. I tried to remove this with:

```
package:
  exclude:
    - ./**
  include:
    - dist/**
    - node_modules/**
    - '!node_modules/aws-sdk/**'
```

But this causes all dev dependencies to be included again, resulting in 30MB! So there seems to be no way to exclude everything and include only real deps and build files again.

**Comment by andythecoderman** [comment]
On a new project using the latest NPM vs. yarn and node 8 vs. typescript I didn't see this issue like I was with typescript/yarn/node 6. Not sure if any of those factors made the difference. 

 What I ended up doing was using gulp to copy my package.json and lock file over to the output directory then running yarn with the --prod flag so it wouldn't install any of the dev dependencies

**Comment by tommedema** [comment]
@andythecoderman I upgraded yarn, npm etc. but still face the same issue.
